### PR TITLE
Add tests for built_in methods support as default field values

### LINKED ISSUE #1857: [Bug]: CLI create API incorrectly requires field value that has been excluded in the config
### What happened?

The DAB CLI hosted API for creating a new record incorrectly requires field value that has been specifically excluded in the config.

### Version

0.8.52

### What database are you using?

Azure SQL

### What hosting model are you using?

Local (including CLI), Static Web Apps (SWA)

### Which API approach are you accessing DAB through?

GraphQL

### Relevant log output

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by abhishekkumams**
## Why make this change?

- Closes #1857 
  - A bug was identified where we were seeing failures with REST/GraphQL mutation queries with tables having default values of columns as methods.
  - It was only happening in MSSQL.

## What is this change?
- The above bug is resolved by the sql client and there are no more failures due to column default value as methods.
- To verify that, I have added Tests for MSSQL, MYSQL, and POSTGRES, that confirms support for methods as default values for entity columns are supported.
- We have added tests to show that all the queries run fine irrespective of columns are present in the excluded fields or not.
- Tests for both REST and GraphQL are added.

## How was this tested?

- [X] Integration Tests
Added MySQL tests are failing because of the bug highlighted [here](#2014).
It's currently marked to ignore in the test suite to not get blocked.

### we have excluded `current_date` and `next_date` for the `create`.
image

image

### CODE REVIEW

**Comment by seantleonard** [comment]
need pr description updated

File: src/Service.Tests/SqlTests/GraphQLMutationTests/GraphQLMutationTestBase.cs [hunk 1]
```diff
@@ -40,6 +40,56 @@ public async Task InsertMutation(string dbQuery)
...
+                        default_function_string_with_parenthesis
+                        default_integer
+                        default_date_string
+                    }
+                }
+            ";
+
+            JsonElement result = await ExecuteGraphQLRequestAsync(graphQLMutation, graphQLMutationName, isAuthenticated: true);
+
+            // Assert the values
```
Comment (Aniruddh25) [inline src/Service.Tests/SqlTests/GraphQLMutationTests/GraphQLMutationTestBase.cs:82]: if you look at other tests, we query the database to find the json object and compare those, why not do the same here?

**Review by Aniruddh25 — CHANGES_REQUESTED** [review]
How does it close #1857? I dont see any tests for a column excluded from config but might still error out in createMutation

File: config-generators/mssql-commands.txt [hunk 2]
```diff
@@ -43,6 +43,7 @@ add MappedBookmarks --config "dab-config.MsSql.json" --source "mappedbookmarks"
 add FteData --config "dab-config.MsSql.json" --source fte_data --permissions "anonymous:*" --graphql "FteData:FteData"
 add InternData --config "dab-config.MsSql.json" --source intern_data --permissions "anonymous:*" --graphql "InternData:InternData"
 add BooksSold --config "dab-config.MsSql.json" --source "books_sold" --rest true --graphql "books_sold:books_sold" --permissions "anonymous:*"
+add DefaultBuiltInFunction --config "dab-config.MsSql.json" --source "default_with_function_table" --rest true --graphql true --permissions "anonymous:*"
```
Comment (seantleonard) [inline config-generators/mssql-commands.txt:46]: as Ani mentioned, you'll need a test to validate the erroneous behavior reported in the issue. 
add permissions that exclude a column that has a default value as a built-in function, and ensure that the request executes as expected.

**Review by seantleonard — CHANGES_REQUESTED** [review]

**Comment by seantleonard** [comment]
@abhishekkumams while we identified a certain behavior for MSSQL, can you elaborate in your pr description how the other database types behave?

File: src/Service.Tests/SqlTests/GraphQLMutationTests/GraphQLMutationTestBase.cs (hunk 1 above)
Comment (abhishekkumams) [reply src/Service.Tests/SqlTests/GraphQLMutationTests/GraphQLMutationTestBase.cs:82]: mutation query already does that and returns the response. other tests we are checking if the value inserted are same as the response returned. we already have other tests conforming it. Here we wanted to tests if the default values are inserted or not. and this test is sufficient.
Furthermore, the value of dates returned by sql query and graphql are not same, so extra step would be required there as well to convert it into same format.

File: src/Service.Tests/dab-config.MsSql.json [hunk 3]
```diff
@@ -2,7 +2,7 @@
   "$schema": "https://github.com/Azure/data-api-builder/releases/download/vmajor.minor.patch/dab.draft.schema.json",
   "data-source": {
     "database-type": "mssql",
-    "connection-string": "Server=tcp:127.0.0.1,1433;Persist Security Info=False;User ID=sa;Password=REPLACEME;MultipleActiveResultSets=False;Connection Timeout=5;",
+    "connection-string": "Server=tcp:127.0.0.1,1433;Persist Security Info=False;User ID=dabuser;Password=Adabpassword@1;MultipleActiveResultSets=False;Connection Timeout=5;TrustServerCertificate=True;",
```
Comment (seantleonard) [inline src/Service.Tests/dab-config.MsSql.json:5]: This change isn't needed, is it?

File: config-generators/mssql-commands.txt [hunk 4]
```diff
@@ -188,3 +189,4 @@ update series --config "dab-config.MsSql.json" --permissions "TestNestedFilterMa
 update series --config "dab-config.MsSql.json" --permissions "TestNestedFilterManyOne_EntityReadForbidden:create,update,delete"
 update series --config "dab-config.MsSql.json" --permissions "TestNestedFilterOneMany_ColumnForbidden:read"
 update series --config "dab-config.MsSql.json" --permissions "TestNestedFilterOneMany_EntityReadForbidden:read"
+update DefaultBuiltInFunction --config "dab-config.MsSql.json" --permissions "anonymous:create" --fields.exclude "current_date,next_date"
```
Comment (seantleonard) [inline config-generators/mssql-commands.txt:192]: Is this a bug? there are no columns in the associated table name `next_date`. I only see `next_day`. Is there an aliasing that occurs somewhere that I missed? Shouldn't DAB error out if you provided an invalid column name?

File: src/Service.Tests/dab-config.MsSql.json [hunk 5]
```diff
@@ -2844,6 +2844,40 @@
...
+          "actions": [
+            {
+              "action": "create",
+              "fields": {
+                  "include": [
+                      "*"
+                  ],
+                  "exclude": [
+                      "current_date",
+                      "next_day"
```
Comment (seantleonard) [inline src/Service.Tests/dab-config.MsSql.json:2870]: this config seems correct, but all the `mssql-commands.txt` files use `next_date`

File: src/Service.Tests/SqlTests/GraphQLMutationTests/GraphQLMutationTestBase.cs [hunk 6]
```diff
@@ -40,6 +40,56 @@ public async Task InsertMutation(string dbQuery)
             SqlTestHelper.PerformTestEqualJsonStrings(expected, actual.ToString());
         }
 
+        /// <summary>
+        /// <code>Do: </code> Inserts new row and return all its columns
+        /// <code>Check: </code> A row is inserted in the table that has rows with default values as built_in methods.
+        /// it should insert it correctly with default values correctly handled by database.
+        /// current_date, current_timestamp, random_number, next_day have default value as built_in methods GETDATE(), NOW(), RAND(), DATEADD(), resp.
+        /// default_string_with_parenthesis has default value "()", default_function_string_with_parenthesis has default value "NOW()".
+        /// default_integer has default value 100, default_date_string has default value "1999-01-08 10:23:54".
```
Comment (seantleonard) [inline src/Service.Tests/SqlTests/GraphQLMutationTests/GraphQLMutationTestBase.cs:49]: need to mention how adding excluded fields to runtime config permissions affects this test and what is expected. Also update pr description describing how the tests address / cover the case outlined in the customer raised issue.

**Review by seantleonard — COMMENTED** [review]
Need some config corrections + elaboration on excluded columns and testing.

File: config-generators/mssql-commands.txt (hunk 4 above)
Comment (abhishekkumams) [reply config-generators/mssql-commands.txt:192]: it just ignores, if a column is mentioned which does not exist.

File: src/Service.Tests/dab-config.MsSql.json (hunk 5 above)
Comment (abhishekkumams) [reply src/Service.Tests/dab-config.MsSql.json:2870]: updated

File: src/Service.Tests/SqlTests/GraphQLMutationTests/GraphQLMutationTestBase.cs (hunk 6 above)
Comment (abhishekkumams) [reply src/Service.Tests/SqlTests/GraphQLMutationTests/GraphQLMutationTestBase.cs]: i will update the summary highlighting that the test captures the scenario of using fields that have default built_in methods irrespective of it being in the list of excluded fields

**Comment by seantleonard** [comment]
Why draft?

**Comment by abhishekkumams** [comment]
> Why draft?

need to add more test before I can open it up for review.

File: src/Service.Tests/SqlTests/GraphQLMutationTests/GraphQLMutationTestBase.cs (hunk 1 above)
Comment (abhishekkumams) [reply src/Service.Tests/SqlTests/GraphQLMutationTests/GraphQLMutationTestBase.cs:82]: i'll update the test to have both checks

File: src/Service.Tests/DatabaseSchema-MySql.sql [hunk 7]
```diff
@@ -221,6 +222,20 @@ CREATE TABLE books_sold (
   last_sold_on_date DATETIME GENERATED ALWAYS AS (last_sold_on) STORED
 );
 
+CREATE TABLE default_with_function_table
+(
+    id INT AUTO_INCREMENT PRIMARY KEY,
+    user_value INT,
+    `current_date` TIMESTAMP DEFAULT (CURRENT_DATE) NOT NULL,
+    `current_timestamp` TIMESTAMP DEFAULT (NOW()) NOT NULL,
```
Comment (severussundar) [inline src/Service.Tests/DatabaseSchema-MySql.sql:230]: Why do we need to enclose current_date and current_timestamp columns within `` and while its not required for other column names? Are these two reserved names?

File: src/Service.Tests/SqlTests/GraphQLMutationTests/GraphQLMutationTestBase.cs (hunk 6 above)
Comment (seantleonard) [reply src/Service.Tests/SqlTests/GraphQLMutationTests/GraphQLMutationTestBase.cs]: The issue was discovered  due to excluding fields that used functions as default values. So need to explain how fixing this issue also fixes the experience/bug encountered when referencing fields that define default values with functions.

**Review by seantleonard — APPROVED** [review]

File: src/Service.Tests/SqlTests/SqlTestHelper.cs [hunk 8]
```diff
@@ -63,12 +63,63 @@ public static bool JsonStringsDeepEqual(string jsonString1, string jsonString2)
...
+
+            JToken expectedJObject = JToken.Parse(expected);
+            JToken actualJObject = JToken.Parse(actual);
+
+            string dateTimeFormat = "yyyy-MM-ddTHH:mm:ss.fffZ"; // ISO 8601 format
+
+            // Function to convert all DateTime values to a consistent format
+            // The convertDateTime function is a local function inside the PerformTestEqualJsonStrings method.
+            // It's used to encapsulate the logic for converting DateTime values to ISO 8601 format.
+            // This makes the PerformTestEqualJsonStrings method easier to read and understand.
```
[the remark this answers is not in the document]
Comment (abhishekkumams) [reply src/Service.Tests/SqlTests/SqlTestHelper.cs:89]: nope, just different formats. As before it was just doing string comparison making "20-09-2004 20:03:50" will be different from "20-09-2004T20:03:50". i'll update the summary with examples.

File: src/Service.Tests/DatabaseSchema-MySql.sql [hunk 9]
```diff
@@ -221,6 +222,20 @@ CREATE TABLE books_sold (
...
+    random_number INT DEFAULT (FLOOR(RAND() * 1000)) NOT NULL,
+    next_date TIMESTAMP DEFAULT (CURRENT_DATE + INTERVAL 1 DAY),
+    default_string_with_parenthesis VARCHAR(100) DEFAULT ('()'),
+    default_function_string_with_parenthesis VARCHAR(100) DEFAULT ('NOW()'),
+    default_integer INT DEFAULT 100,
+    default_date_string DATETIME DEFAULT ("1999-01-08 10:23:54")
+);
```
Comment (Aniruddh25) [inline src/Service.Tests/DatabaseSchema-MySql.sql:237]:
```suggestion
);```

File: src/Service.Tests/SqlTests/GraphQLMutationTests/PostgreSqlGraphQLMutationTests.cs [hunk 10]
```diff
@@ -112,6 +112,28 @@ public async Task InsertMutationForComputedColumns()
             await InsertMutationForComputedColumns(postgresQuery);
         }
 
+        /// <summary>
+        /// <code>Do: </code> Inserts new row in a table containing default values as built_in methods.
+        /// <code>Check: </code> Correctly inserts the row with columns having default values as built_in methods and returns the inserted row
+        /// as graphQL response.
```
Comment (Aniruddh25) [inline src/Service.Tests/SqlTests/GraphQLMutationTests/PostgreSqlGraphQLMutationTests.cs:118]:
```suggestion
        /// as GraphQL response.
```

File: src/Service.Tests/SqlTests/GraphQLMutationTests/GraphQLMutationTestBase.cs [hunk 11]
```diff
@@ -40,6 +40,62 @@ public async Task InsertMutation(string dbQuery)
             SqlTestHelper.PerformTestEqualJsonStrings(expected, actual.ToString());
         }
 
+        /// <summary>
+        /// <code>Do: </code> Inserts new row and return all its columns
+        /// <code>Check: </code> A row is inserted in the table that has rows with default values as built_in methods.
+        /// it should insert it correctly with default values correctly handled by database.
```
Comment (Aniruddh25) [inline src/Service.Tests/SqlTests/GraphQLMutationTests/GraphQLMutationTestBase.cs:46]:
```suggestion
        /// it should insert it correctly with default values handled by database.
```

File: src/Service.Tests/SqlTests/GraphQLMutationTests/MsSqlGraphQLMutationTests.cs [hunk 12]
```diff
@@ -58,6 +58,27 @@ public async Task InsertMutation()
             await InsertMutation(msSqlQuery);
         }
 
+        /// <summary>
+        /// <code>Do: </code> Inserts new row in a table containing default values as built_in methods.
+        /// <code>Check: </code> Correctly inserts the row with columns having default values as built_in methods and returns the inserted row
+        /// as graphQL response.
```
Comment (Aniruddh25) [inline src/Service.Tests/SqlTests/GraphQLMutationTests/MsSqlGraphQLMutationTests.cs:64]:
```suggestion
        /// as GraphQL response.
```

**Review by Aniruddh25 — APPROVED** [review]
LGTM, Left few nits

File: src/Service.Tests/DatabaseSchema-MySql.sql [hunk 13]
```diff
@@ -221,6 +222,20 @@ CREATE TABLE books_sold (
...
+
```
Comment (Aniruddh25) [inline src/Service.Tests/DatabaseSchema-MySql.sql:238]:
```suggestion
```

File: src/Service.Tests/SqlTests/GraphQLMutationTests/MySqlGraphQLMutationTests.cs [hunk 14]
```diff
@@ -96,6 +96,19 @@ public async Task InsertMutationForComputedColumns()
             await InsertMutationForComputedColumns(mySqlQuery);
         }
 
+        /// <summary>
+        /// <code>Do: </code> Inserts new row in a table containing default values as built_in methods.
+        /// <code>Check: </code> Correctly inserts the row with columns having default values as built_in methods and returns the inserted row
+        /// as graphQL response.
```
Comment (Aniruddh25) [inline src/Service.Tests/SqlTests/GraphQLMutationTests/MySqlGraphQLMutationTests.cs:102]:
```suggestion
        /// as GraphQL response.
```

**Pull request merged**